### PR TITLE
added week4 homework JS1

### PR DESCRIPTION
**Pull request by ajgcph2019**

### CODE REVIEW

File: javascript/javascript1/week4/characterfrequency.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+function getCharacterFrequencies(incomingWord) {
+  const result = {};
+
+  for (let i = 0; i < incomingWord.length; i++) {
+    let character = incomingWord.charAt(i);
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/characterfrequency.js:5]: incomingWord[i] is clearer

File: javascript/javascript1/week4/characterfrequency.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+      result[character]++;
+    } else {
+      result[character] = 1;
+    }
+  }
+
+  return result;
+}
+
+const charObject = getCharacterFrequencies("happy");
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/characterfrequency.js:19]: looks good

File: javascript/javascript1/week4/creditcardtype.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+   //Discover
+   discoverRegex = new RegExp('^(6011|65|64[4-9]|62212[6-9]|6221[3-9]|622[2-8]|6229[01]|62292[0-5])[0-9]{0,}$');
+   ////6011, 622126-622925, 644-649, 65
+
+  console.log(accountNumber);
+
+
+   let cardType = "unknown";
+   if (accountNumber.match(jcbRegex)) {
+       cardType = "jcb";
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/creditcardtype.js:23]: you can do `return "jcb"` and similar everywhere else.

File: javascript/javascript1/week4/voiceassistant.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
+let name = '';
+let todos = [];
+let favoriteDish = [];
+let events = [];
+var months = ["January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December"];
+const arrayOfCommands = [
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/voiceassistant.js:6]: i didn't get into details, but looks good

File: javascript/javascript1/week4/palindrome.js [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        // we found an occurance that matches the current letter
+        firstOccurence = j;
+
+        // generate the substring to match
+        let subString = incomingString.slice(i, firstOccurence + 1);
+
+        // check if the substring is a palindrome and if it is longer then longest
+        if (
+          subString.length > longest.length &&
+          subString.toLowerCase() ===
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/palindrome.js:21]: nice

File: javascript/javascript1/week4/palindrome.js [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          longest = subString;
+        }
+      }
+    }
+  }
+  // return the longest palindrome
+  return longest;
+}
+
+const str = isPalindrome("My dad is a racecar athlete");
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/palindrome.js:37]: this type of problem is typical for a programming test. If you have a good understanding of this, it's good preparation for job interviews etc.

File: javascript/javascript1/week4/creditcardformatter.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+function formatCreditCardNumber(number) {
+  if (typeof number == "string") {
+    console.log("Enter a valid card number");
+  } 
+  else {
+     number = number.toString();
+     const spaceBetweenDigits = 4;
+     let result = " ";
+     while (number.length > 0) {
```
Comment (aleksd-dfds) [inline javascript/javascript1/week4/creditcardformatter.js:9]: i think you can also do this simpler by using string.join().

File: javascript/javascript1/week4/characterfrequency.js (hunk 1 above)
Comment (ajgcph2019) [reply javascript/javascript1/week4/characterfrequency.js:5]: function getCharacterFrequencies(incomingWord) {
  const result = {};

  for (let i = 0; i < incomingWord.length; i++) {
    let character = incomingWord[i];

    console.log(character);

    if (result[character]) {
      result[character]++;
    } else {
      result[character] = 1;
    }
  }

  return result;
}

const charObject = getCharacterFrequencies("happy");
console.log(charObject);

Comment (ajgcph2019) [reply javascript/javascript1/week4/characterfrequency.js:5]: corrected..

File: javascript/javascript1/week4/creditcardtype.js (hunk 3 above)
Comment (ajgcph2019) [reply javascript/javascript1/week4/creditcardtype.js:23]: function getCreditCardType(accountNumber)
{
   //JCB
   jcbRegex = new RegExp('^(?:2131|1800|35)[0-9]{0,}$'); //2131, 1800, 35 (3528-3589)
   // American Express
  amexRegex = new RegExp('^3[47][0-9]{0,}$'); //34, 37
   // Diners Club
   dinersRegex = new RegExp(/^3(?:0[0-59]{1}|[689])[0-9]{0,}$/); //300-305, 309, 36, 38-39
   // Visa
  visaRegex = new RegExp('^4[0-9]{0,}$'); //4
   // MasterCard
   mastercardRegex = new RegExp('^(5[1-5]|222[1-9]|22[3-9]|2[3-6]|27[01]|2720)[0-9]{0,}$'); //2221-2720, 51-55
   maestroRegex = new RegExp('^(5[06789]|6)[0-9]{0,}$'); //always growing in the range: 60-69, started with / not something else, but starting 5 must be encoded as mastercard anyway
   //Discover
   cdiscoverRegex = new RegExp('^(6011|65|64[4-9]|62212[6-9]|6221[3-9]|622[2-8]|6229[01]|62292[0-5])[0-9]{0,}$');
   ////6011, 622126-622925, 644-649, 65
  console.log(accountNumber);
  const cardType = "";
   if (accountNumber.match(jcbRegex)) {
       return "jcb";
   } else if (accountNumber.match(amexRegex)) {
    return "amex";
   } else if (accountNumber.match(dinersRegex)) {
    return "diners_club";
   } else if (accountNumber.match(visaRegex)) {
       return "visa";
   } else if (accountNumber.match(mastercardRegex)) {
       return "mastercard";
   } else if (accountNumber.match(discoverRegex)) {
       return "discover";
   } else if (accountNumber.match(maestroRegex)) {
       if (accountNumber[0] == '5') { //started 5 must be mastercard
           return "mastercard";
       } else {
           return "maestro"; //maestro is all 60-69 which is not something else, thats why this condition in the end
       }
   }
   return cardType;
}
const cardInfo = getCreditCardType('4781321334789876');
console.log(cardInfo);

File: javascript/javascript1/week4/characterfrequency.js (hunk 2 above)
Comment (ajgcph2019) [reply javascript/javascript1/week4/characterfrequency.js:19]: function getCreditCardType(accountNumber) {
  //JCB
  jcbRegex = new RegExp("^(?:2131|1800|35)[0-9]{0,}$"); //2131, 1800, 35 (3528-3589)
  // American Express
  amexRegex = new RegExp("^3[47][0-9]{0,}$"); //34, 37
  // Diners Club
  dinersRegex = new RegExp(/^3(?:0[0-59]{1}|[689])[0-9]{0,}$/); //300-305, 309, 36, 38-39
  // Visa
  visaRegex = new RegExp("^4[0-9]{0,}$"); //4
  // MasterCard
  mastercardRegex = new RegExp(
    "^(5[1-5]|222[1-9]|22[3-9]|2[3-6]|27[01]|2720)[0-9]{0,}$"
  ); //2221-2720, 51-55
  maestroRegex = new RegExp("^(5[06789]|6)[0-9]{0,}$"); //always growing in the range: 60-69, started with / not something else, but starting 5 must be encoded as mastercard anyway
  //Discover
  cdiscoverRegex = new RegExp(
    "^(6011|65|64[4-9]|62212[6-9]|6221[3-9]|622[2-8]|6229[01]|62292[0-5])[0-9]{0,}$"
  );
  ////6011, 622126-622925, 644-649, 65
  console.log(accountNumber);
  const cardType = "";
  if (accountNumber.match(jcbRegex)) {
    return "jcb";
  } else if (accountNumber.match(amexRegex)) {
    return "amex";
  } else if (accountNumber.match(dinersRegex)) {
    return "diners_club";
  } else if (accountNumber.match(visaRegex)) {
    return "visa";
  } else if (accountNumber.match(mastercardRegex)) {
    return "mastercard";
  } else if (accountNumber.match(discoverRegex)) {
    return "discover";
  } else if (accountNumber.match(maestroRegex)) {
    if (accountNumber[0] == "5") {
      //started 5 must be mastercard
      return "mastercard";
    } else {
      return "maestro"; //maestro is all 60-69 which is not something else, thats why this condition in the end
    }
  }
  return cardType;
}
const cardInfo = getCreditCardType("4781321334789876");
console.log(cardInfo);

File: javascript/javascript1/week4/creditcardtype.js (hunk 3 above)
Comment (ajgcph2019) [reply javascript/javascript1/week4/creditcardtype.js:23]: function getCreditCardType(accountNumber) {
  //JCB
  jcbRegex = new RegExp("^(?:2131|1800|35)[0-9]{0,}$"); //2131, 1800, 35 (3528-3589)
  // American Express
  amexRegex = new RegExp("^3[47][0-9]{0,}$"); //34, 37
  // Diners Club
  dinersRegex = new RegExp(/^3(?:0[0-59]{1}|[689])[0-9]{0,}$/); //300-305, 309, 36, 38-39
  // Visa
  visaRegex = new RegExp("^4[0-9]{0,}$"); //4
  // MasterCard
  mastercardRegex = new RegExp(
    "^(5[1-5]|222[1-9]|22[3-9]|2[3-6]|27[01]|2720)[0-9]{0,}$"
  ); //2221-2720, 51-55
  maestroRegex = new RegExp("^(5[06789]|6)[0-9]{0,}$"); //always growing in the range: 60-69, started with / not something else, but starting 5 must be encoded as mastercard anyway
  //Discover
  cdiscoverRegex = new RegExp(
    "^(6011|65|64[4-9]|62212[6-9]|6221[3-9]|622[2-8]|6229[01]|62292[0-5])[0-9]{0,}$"
  );
  ////6011, 622126-622925, 644-649, 65
  console.log(accountNumber);
  const cardType = "";
  if (accountNumber.match(jcbRegex)) {
    return "jcb";
  } else if (accountNumber.match(amexRegex)) {
    return "amex";
  } else if (accountNumber.match(dinersRegex)) {
    return "diners_club";
  } else if (accountNumber.match(visaRegex)) {
    return "visa";
  } else if (accountNumber.match(mastercardRegex)) {
    return "mastercard";
  } else if (accountNumber.match(discoverRegex)) {
    return "discover";
  } else if (accountNumber.match(maestroRegex)) {
    if (accountNumber[0] == "5") {
      //started 5 must be mastercard
      return "mastercard";
    } else {
      return "maestro"; //maestro is all 60-69 which is not something else, thats why this condition in the end
    }
  }
  return cardType;
}
const cardInfo = getCreditCardType("4781321334789876");
console.log(cardInfo);

File: javascript/javascript1/week4/palindrome.js (hunk 6 above)
Comment (ajgcph2019) [reply javascript/javascript1/week4/palindrome.js:37]: Sure yes thank you for reminding that these types of questions are interview questions

File: javascript/javascript1/week4/creditcardformatter.js (hunk 7 above)
Comment (ajgcph2019) [reply javascript/javascript1/week4/creditcardformatter.js:9]: function formatCreditCardNumber(number) {
    const string = number.toString();
    const array = [];
    for (let i = 0; i < string.length; i += 4) {
        array.push(string.slice(i, i + 4));
    }
    return array.join(' ');
}

const formattedCreditCardObject = formatCreditCardNumber(123456789);
console.log(formattedCreditCardObject);

**Pull request merged**